mist-core/src/main/java/edu/snu/mist/core/task/recovery/TaskSideRecoveryStarter.java

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+package edu.snu.mist.core.task.recovery;
+
+import edu.snu.mist.core.task.checkpointing.CheckpointManager;
+import edu.snu.mist.formats.avro.RecoveryInfo;
+import edu.snu.mist.formats.avro.TaskToMasterMessage;
+import org.apache.avro.AvroRemoteException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The runnable class for recovering queries leveraging a single thread.
+ */
+public class TaskSideRecoveryStarter implements Runnable {
+
+  private static final Logger LOG = Logger.getLogger(TaskSideRecoveryStarter.class.getName());
+
+  /**
+   * Indicates whether the recovery is running or not.
+   */
+  private AtomicBoolean isRecoveryRunning;
+
+  /**
+   * The proxy for avro rpc from task to master.
+   */
+  private TaskToMasterMessage proxyToMaster;
+
+  /**
+   * The executor service used for group recovery.
+   */
+  private ExecutorService executorService;
+
+  /**
+   * The checkpoint manager for loading checkpoints.
+   */
+  private CheckpointManager checkpointManager;
+
+  /**
+   * The task hostname which can be seen from MistMaster.
+   */
+  private String taskHostname;
+
+  public TaskSideRecoveryStarter(final AtomicBoolean isRecoveryRunning,
+                                 final TaskToMasterMessage proxyToMaster,
+                                 final CheckpointManager checkpointManager,
+                                 final String taskHostname,
+                                 final int numTheads) {
+    this.isRecoveryRunning = isRecoveryRunning;
+    this.proxyToMaster = proxyToMaster;
+    this.checkpointManager = checkpointManager;
+    this.taskHostname = taskHostname;
+    this.executorService = Executors.newFixedThreadPool(numTheads);
+  }
+
+  public void run() {
+    // Get the recovery info from the MistMaster.
+    try {
+      while (true) {
+        final RecoveryInfo recoveryInfo = proxyToMaster.pullRecoveringGroups(taskHostname);
+        LOG.log(Level.INFO, "Recovering groups: {0}", recoveryInfo.getRecoveryGroupList().toString());
+        if (recoveryInfo.getRecoveryGroupList().isEmpty()) {
+          // Notify that recovery is done!
+          isRecoveryRunning.set(false);
+          // Finish the runner thread.
+          break;
+        } else {
+          final List<Future> futureList = new ArrayList<>();
+          for (final String recoveryGroup : recoveryInfo.getRecoveryGroupList()) {
+            futureList.add(executorService.submit(new RecoveryRunner(recoveryGroup, checkpointManager)));
+          }
+          // Wait for the all group recovery finishes.
+          for (final Future future : futureList) {
+            future.get();
```

Comment:
Why do you wait?

Reply:
This is to prevent two or more concurrent recoveries are done in the same task.

Reply:
How to prevent if this runnable called twice?

Reply:
This is prevented using atomic variable. See `startRecovery` in `DefaultRecoveryManagerImpl`.